include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -250,118 +257,341 @@ __parallel_for(_ExecutionPolicy&& __exec, _Fp __brick, _Index __count, _Ranges&&
+        auto __res_acc = __res.template get_access<access_mode::write>(__cgh);
+        __cgh.single_task<_ReduceKernel>([=] {
+            _Tp __result;
+            __transform_pattern(__n, __result, __rngs...);
+            __reduce_pattern.apply_init(__init, __result);
+            __res_acc[0] = __result;
+        });
+    });
 
-    auto __max_compute_units = oneapi::dpl::__internal::__max_compute_units(__exec);
-    // TODO: find a way to generalize getting of reliable work-group size
-    ::std::size_t __work_group_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
-    // change __work_group_size according to local memory limit
-    __work_group_size = oneapi::dpl::__internal::__max_local_allocation_size(::std::forward<_ExecutionPolicy>(__exec),
-                                                                             sizeof(_Tp), __work_group_size);
+    return __future(__reduce_event, __res);
+}
 
-#if _ONEDPL_COMPILE_KERNEL
-    auto __kernel = __internal::__kernel_compiler<_ReduceKernel>::__compile(::std::forward<_ExecutionPolicy>(__exec));
-    __work_group_size = ::std::min(__work_group_size, oneapi::dpl::__internal::__kernel_work_group_size(
-                                                          ::std::forward<_ExecutionPolicy>(__exec), __kernel));
-#endif
-    ::std::size_t __iters_per_work_item = __grainsize;
-    // distribution is ~1 work groups per compute init
-    if (__exec.queue().get_device().is_cpu())
-        __iters_per_work_item = (__n - 1) / (__max_compute_units * __work_group_size) + 1;
-    // __iters_per_work_item shows number of elements to reduce on global memory
-    // __work_group_size shows number of elements to reduce on local memory
-    // guarantee convergence of the reduction
-    if (__iters_per_work_item == 1 && __work_group_size == 1)
-        __iters_per_work_item = 2;
-
-    ::std::size_t __size_per_work_group =
-        __iters_per_work_item * __work_group_size;            // number of buffer elements processed within workgroup
-    _Size __n_groups = (__n - 1) / __size_per_work_group + 1; // number of work groups
-    _Size __n_items = (__n - 1) / __iters_per_work_item + 1;  // number of work items
-
-    _PRINT_INFO_IN_DEBUG_MODE(__exec, __work_group_size, __max_compute_units);
-
-    // Create temporary global buffers to store temporary values
-    sycl::buffer<_Tp> __temp(sycl::range<1>(2 * __n_groups));
-    // __is_first == true. Reduce over each work_group
-    // __is_first == false. Reduce between work groups
-    bool __is_first = true;
-
-    // For memory utilization it's better to use one big buffer instead of two small because size of the buffer is close to a few MB
-    _Size __offset_1 = 0;
-    _Size __offset_2 = __n_groups;
-
-    sycl::event __reduce_event;
-    do
+// Parallel_transform_reduce for a single work group
+template <::std::uint16_t __work_group_size, ::std::size_t __iters_per_work_item, typename _Tp, typename _ReduceOp,
```

Comment:
Here we have to pass `::std::uint16_t __work_group_size`, and `::std::size_t __iters_per_work_item` in compile time as the template parameters. At the same time, a caller calculates these parameters in runtime..
Furthermore,   it seems that structure doesn't use any "compile time specific" - there is `constexpr` expressions and C style private arrays....
So, probably we may use runtime variables and reduce number of kernels.

Reply:
This enables the usage of `transform_init_known` with a compile-time `__iters_per_work_item` parameter for better memory optimization (c.f., the declaration of `__transform_pattern` in line 317).